### PR TITLE
Update Codebase to use React 16 - DO NOT MERGE

### LINKED ISSUE #658: React 16 upgrade
# Issue Description
When react 16 comes out, we'll want to enable compatibility with react 16.

## Issue Type

- [ ] New Feature
- [x] Enhancement
- [ ] Bug
- [ ] Other

## Expected Behavior
Terra components are compatible with React 16

## Current Behavior
N/A

### PR DESCRIPTION
**Pull request by JakeLaCombe**
Resolves #658 

Testing Branch for SD's to test terra core against.

You can find React 16 here: https://terra-core-deployed-pr-1175.herokuapp.com/static/#/site

ghfilter
- .*\.md

### CODE REVIEW

**Comment by StephenEsser** [comment]
image

**Comment by JakeLaCombe** [comment]
@StephenEsser more to come as well! Almost all the wdio snapshots need to be updated :O

**Comment by mhemesath** [comment]
We need to fix all the screenshots that were just on the element and now are full page. Those waste size and less sensitive to visual diffs.

**Comment by bjankord** [comment]
@mhemesath Yeah I'm gonna sync with Jake with he's back in the office to get these updated.

**Comment by bjankord** [comment]
Seeing a few console errors on different pages.

* https://terra-core-deployed-pr-1175.herokuapp.com/static/#/site/date-time-picker
<details>
 <summary>Error</summary>

```
warning.js:33 Warning: Failed prop type: Invalid prop `selected` of type `string` supplied to `DatePicker`, expected `object`.
```
</details>


* https://terra-core-deployed-pr-1175.herokuapp.com/static/#/site/embedded-content-consumer
<details>
 <summary>Error</summary>

```
warning.js:33 Warning: Received `true` for a non-boolean attribute `fill`.
If you want to write it to the DOM, pass a string instead: fill="true" or fill={value.toString()}.
```
</details>


* https://terra-core-deployed-pr-1175.herokuapp.com/static/#/site/form/input
* https://terra-core-deployed-pr-1175.herokuapp.com/static/#/site/form/number-field
<details>
 <summary>Error</summary>

```
warning.js:33 Warning: Each child in an array or iterator should have a unique "key" prop.

Check the render method of `PropsTable`. See https://fb.me/react-warning-keys for more information.
```
</details>



* https://terra-core-deployed-pr-1175.herokuapp.com/static/#/site/form-radio
<details>
 <summary>Error</summary>

```
Warning: React does not recognize the `isDisabled` prop on a DOM element. 
If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase 
`isdisabled` instead. If you accidentally passed it from a parent component, remove it from the 
DOM element.
```
</details>


* https://terra-core-deployed-pr-1175.herokuapp.com/static/#/site/image
<details>
 <summary>Error</summary>

```
warning.js:33 Warning: React does not recognize the `placeholderSrc` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `placeholdersrc` instead. If you accidentally passed it from a parent component, remove it from the DOM element.
```
</details>

**Comment by bjankord** [comment]
Need to look into the branch deployment for this. It is currently failing out when trying to run webpack. It looks like its calling webpack before maybe npm install has run?

```
 > webpack --config webpack.prod.config --progress
       
module.js:538
    throw err;
    ^
Error: Cannot find module 'postcss-assets-webpack-plugin'
```

File: packages/terra-date-picker/tests/nightwatch/components/DatePickerDefault.jsx [hunk 1]
```diff
@@ -6,9 +6,301 @@ const locale = document.getElementsByTagName('html')[0].getAttribute('lang');
 
 const DatePickerDefault = () => (
   <Base locale={locale}>
-    <DatePicker
-      name="date-input"
-    />
+    <div>
```
Comment (emilyrohrbough) [inline packages/terra-date-picker/tests/nightwatch/components/DatePickerDefault.jsx]: Why is this like this?

File: packages/terra-date-picker/tests/nightwatch/date-picker-spec.js [hunk 2]
```diff
@@ -139,7 +139,7 @@ module.exports = resizeTo(['tiny', 'huge'], {
 
     browser.expect.element('.react-datepicker').to.be.present;
 
-    browser.click('.react-datepicker__today-button');
+    browser.click('[aria-label="day-10"]');
```
Comment (emilyrohrbough) [inline packages/terra-date-picker/tests/nightwatch/date-picker-spec.js]: Is this mapped to the same button?

File: packages/terra-demographics-banner/src/DemographicsBanner.jsx [hunk 3]
```diff
@@ -67,6 +67,7 @@ const defaultProps = {
 const contextTypes = {
   /* eslint-disable consistent-return */
   intl: (context) => {
+    console.log('Intl Checking');
```
Comment (emilyrohrbough) [inline packages/terra-demographics-banner/src/DemographicsBanner.jsx]: Needs removed

File: packages/terra-demographics-banner/tests/jest/DemographicsBanner.test.jsx [hunk 4]
```diff
@@ -8,12 +8,12 @@ import messages from '../../translations/en-US.json';
...
 const locale = 'en-US';
 
-it('renders a blank banner wrapper', () => {
+xit('renders a blank banner wrapper', () => {
   const banner = shallow(<IntlProvider locale={locale} messages={messages}><DemographicsBanner /></IntlProvider>);
   expect(banner).toMatchSnapshot();
 });
 
-it('renders the banner wrapper with all props', () => {
+xit('renders the banner wrapper with all props', () => {
```
Comment (emilyrohrbough) [inline packages/terra-demographics-banner/tests/jest/DemographicsBanner.test.jsx]: oops typo

File: packages/terra-demographics-banner/tests/jest/DemographicsBanner.test.jsx [hunk 5]
```diff
@@ -8,12 +8,12 @@ import messages from '../../translations/en-US.json';
 
 const locale = 'en-US';
 
-it('renders a blank banner wrapper', () => {
+xit('renders a blank banner wrapper', () => {
```
Comment (emilyrohrbough) [inline packages/terra-demographics-banner/tests/jest/DemographicsBanner.test.jsx]: oops typo

File: packages/terra-heading/tests/jest/Heading.test.jsx [hunk 6]
```diff
@@ -51,12 +51,12 @@ describe('Heading', () => {
   // Prop Tests
   it('should have all props including customProps set correctly', () => {
     const heading = shallow(<Heading level={1} id="id" size="small" color="#f00" weight={200} isItalic isVisuallyHidden>All props and custom attrs</Heading>);
```
Comment (emilyrohrbough) [inline packages/terra-heading/tests/jest/Heading.test.jsx]: Could we add a snapshot test here instead? All of this could be removed then

File: packages/terra-hookshot/tests/jest/Hookshot.test.jsx [hunk 7]
```diff
@@ -7,7 +7,7 @@ it('should shallow a default component', () => {
   expect(hookshot).toMatchSnapshot();
 });
 
-it('should render a default component', () => {
+xit('should render a default component', () => {
```
Comment (emilyrohrbough) [inline packages/terra-hookshot/tests/jest/Hookshot.test.jsx]: oops typo

File: packages/terra-icon/src/IconBase.jsx [hunk 8]
```diff
@@ -81,7 +81,7 @@ const IconBase = ({
 
   attributes.height = height;
   attributes.width = width;
-  attributes.focusable = focusable;
+  attributes.focusable = focusable.toString();
```
Comment (emilyrohrbough) [inline packages/terra-icon/src/IconBase.jsx]: What did we make this change?

File: packages/terra-legacy-theme/CHANGELOG.md [hunk 9]
```diff
@@ -3,6 +3,8 @@ Changelog
 
 Unreleased
 ----------
+### Changed
```
Comment (emilyrohrbough) [inline packages/terra-legacy-theme/CHANGELOG.md]: This isn't needed

File: packages/terra-mixins/CHANGELOG.md [hunk 10]
```diff
@@ -3,6 +3,8 @@ Changelog
 
 Unreleased
 ----------
+### Changed
```
Comment (emilyrohrbough) [inline packages/terra-mixins/CHANGELOG.md]: This isn't needed

File: packages/terra-modal/tests/jest/Modal.test.jsx [hunk 11]
```diff
@@ -11,7 +11,7 @@ it('should mount an open modal', () => {
   expect(modal).toMatchSnapshot();
 });
 
-it('should render an open modal', () => {
+xit('should render an open modal', () => {
```
Comment (emilyrohrbough) [inline packages/terra-modal/tests/jest/Modal.test.jsx]: uncomment

File: packages/terra-popup/tests/jest/Popup.test.jsx [hunk 12]
```diff
@@ -7,7 +7,7 @@ it('should shallow a default component', () => {
   expect(popup).toMatchSnapshot();
 });
 
-it('should render a default component', () => {
+xit('should render a default component', () => {
```
Comment (emilyrohrbough) [inline packages/terra-popup/tests/jest/Popup.test.jsx]: uncomment

File: packages/terra-popup/tests/nightwatch/components/OverlayPopup.jsx [hunk 13]
```diff
@@ -15,7 +15,6 @@ class DefaultPopup extends React.Component {
     this.overlayStyle = document.documentElement.style.overflow;
     this.overlayId = document.documentElement.id;
     document.documentElement.id = 'popup-overlay-test';
-    document.documentElement.style.overflow = 'auto';
```
Comment (emilyrohrbough) [inline packages/terra-popup/tests/nightwatch/components/OverlayPopup.jsx]: Why was this removed?

File: packages/terra-site/package.json [hunk 14]
```diff
@@ -114,8 +114,9 @@
     "postcss-loader": "^2.0.6",
     "postcss-rtl": "^1.1.2",
     "raw-loader": "^0.5.1",
-    "react": "^15.4.2",
-    "react-dom": "^15.4.2",
+    "raf": "^3.4.0",
```
Comment (emilyrohrbough) [inline packages/terra-site/package.json]: What is raf?

File: packages/terra-site/webpack.config.js [hunk 15]
```diff
@@ -23,6 +23,7 @@ const threadLoaderRule = {
 
 module.exports = {
   entry: {
+    raf: 'raf/polyfill',
```
Comment (emilyrohrbough) [inline packages/terra-site/webpack.config.js]: Why are we using two polyfills now? Will this be required by all consumers?

File: packages/terra-status/tests/jest/Status.test.jsx [hunk 16]
```diff
@@ -38,9 +38,9 @@ it('should render a arrange with status and customProps', () => {
 it('should have all props including customProps set correctly', () => {
   const arrangeWithStatus = <Status color="green" id="id" >{arrange}</Status>;
   const wrapper = shallow(arrangeWithStatus);
```
Comment (emilyrohrbough) [inline packages/terra-status/tests/jest/Status.test.jsx]: I think we can update this to a screenshot & jest will do this comparison for us

File: packages/terra-time-input/tests/nightwatch/component/TimeInputTwelveHourMobile.jsx [hunk 17]
```diff
@@ -18,12 +18,6 @@ export default class TimeInputElement extends React.Component {
     }
   }
 
-  componentWillUnmount() {
```
Comment (emilyrohrbough) [inline packages/terra-time-input/tests/nightwatch/component/TimeInputTwelveHourMobile.jsx]: Is this a 16 change? Should this be moved else where?

File: packages/terra-toggle/tests/jest/Toggle.test.jsx [hunk 18]
```diff
@@ -21,9 +21,11 @@ describe('Toggle', () => {
   // Prop Tests
   it('should have all props set correctly', () => {
     const toggle = shallow(<Toggle isAnimated isOpen>All props</Toggle>);
```
Comment (emilyrohrbough) [inline packages/terra-toggle/tests/jest/Toggle.test.jsx]: I think we can update this to a screenshot & jest will do this comparison for us

**Comment by emilyrohrbough** [comment]
We shouldn't need to check in the DEPENDENCIES.md files. That'll get updated during [`postpublish`](https://github.com/cerner/terra-core/blob/master/package.json#L65)

**Comment by JakeLaCombe** [comment]
I'll close this soon and put out a more throughouh review. I wanted a link for functional testing, and also wanted to test the Travis CI builds. I'll be sure to address the comments in this review.

File: packages/terra-site/webpack.config.js (hunk 15 above)
Comment (JakeLaCombe) [reply packages/terra-site/webpack.config.js]: React 16 requires request animation frame. https://reactjs.org/blog/2017/09/26/react-v16.0.html

File: packages/terra-date-picker/tests/nightwatch/components/DatePickerDefault.jsx (hunk 1 above)
Comment (JakeLaCombe) [reply packages/terra-date-picker/tests/nightwatch/components/DatePickerDefault.jsx]: Totally forgot about this!

I had issues using popper, and so I created a large scroll page to test scrolling with popper. This can all be undone.

**Comment by JakeLaCombe** [comment]
Closing. I think I have enough info to verify my next pull request will operate. I'll make sure all comments here are addressed.
